Fix Rails environment when running tests with Ruby

tenderlove opened the pull request:
I frequently run tests with `ruby`, not with a runner like `rake` or
`rails`.  When running the test with just `ruby` the `RAILS_ENV`
environment variable did not get set to "test", and this would cause the
tests to fail (and even mutate the development database!)

This commit adds integration tests for running tests with just `ruby`
and ensures the environment gets defaulted to "test".  I also added a
test to ensure that passing an environment to `-e` actually works (and
fixed that case too).

An interesting / annoying thing is that Minitest picks up it's plugins
by asking RubyGems for a list of files:

  https://github.com/seattlerb/minitest/blob/ca6a71ca901016db09a5ad466b4adea4b52a504a/lib/minitest.rb#L92-L100

This means that RubyGems needs to somehow know about the file before it
can return it to Minitest.  Since we are not packaging Rails as a Gem
before running the integration tests on it (duh, why would you do
that?), RubyGems doesn't know about the file, so it can't tell Minitest,
so Minitest doesn't automatically require it.  This means I had to
manually require and insert the plugin in our integration test.  I've
left comments about that in the test as well.

Ugh.

tenderlove committed:
Fix Rails environment when running tests with Ruby

I frequently run tests with `ruby`, not with a runner like `rake` or
`rails`.  When running the test with just `ruby` the `RAILS_ENV`
environment variable did not get set to "test", and this would cause the
tests to fail (and even mutate the development database!)

This commit adds integration tests for running tests with just `ruby`
and ensures the environment gets defaulted to "test".  I also added a
test to ensure that passing an environment to `-e` actually works (and
fixed that case too).

An interesting / annoying thing is that Minitest picks up it's plugins
by asking RubyGems for a list of files:

  https://github.com/seattlerb/minitest/blob/ca6a71ca901016db09a5ad466b4adea4b52a504a/lib/minitest.rb#L92-L100

This means that RubyGems needs to somehow know about the file before it
can return it to Minitest.  Since we are not packaging Rails as a Gem
before running the integration tests on it (duh, why would you do
that?), RubyGems doesn't know about the file, so it can't tell Minitest,
so Minitest doesn't automatically require it.  This means I had to
manually require and insert the plugin in our integration test.  I've
left comments about that in the test as well.

Ugh.

rafaelfranca commented on railties/test/application/test_runner_test.rb:
```diff
@@ -727,6 +747,42 @@ def create_schema
+                             time: result.time)
+            end
+          end
+          
```
There are trailing whitespaces here.

tenderlove replied:
Thanks!  I saw the complaints, but Vim didn't show for some reason.  😀

rafaelfranca commented on railties/test/application/test_runner_test.rb:
```diff
@@ -727,6 +747,42 @@ def create_schema
+            Minitest.reporter.reporters.clear
+            Minitest.reporter.reporters << JSONReporter.new
+          end
+          
```
There are trailing whitespaces here.

rafaelfranca reviewed (approved):
Oh. That explains it!

eileencodes reviewed (changes requested):
I approved this and then actually tested it (sorry 😞 ) and it's didn't fix the issue on my repo. I tried creating a new one in case I messed something up and the ENV when running with Ruby is still development and not test.

Edouard-chin commented:
Ah doesn't work for me either :(. I think it's already too late to set the ENV side the runner when using ruby or even rake because we already loaded the gems `Bundler.require(*Rails.groups)` (Rails.groups, make a call to Rails.env which memoize 'development')

I think we can add `ENV["RAILS_ENV"] || 'test'` after this condition 🤷‍♂️ https://github.com/rails/rails/blob/bbacd60048a8efa1777a01292a9392e146a7d885/railties/lib/rails/test_unit/railtie.rb#L5-L7

tenderlove commented:
Ah, so this patch is setting the ENV correctly, it's just too late. *sigh*

@Edouard-chin I think we used to have something like that.  We really need a test for this case though because the behavior breaks too often for my liking 😉

tenderlove committed:
Add failing test for wrong database connection

When tests are run with just `ruby`, the RAILS_ENV is set to
`development` too early, and we connect to the development database
rather than the test database.

tenderlove committed:
Revert "remove unnecessary `RAILS_ENV` setting"

This reverts commit 9a80f52541ed2c93ebef02909ecab3aaf9127150.

eileencodes reviewed (approved):
:+1: looks good to me

tenderlove committed:
Set the Rails environment from an environment variable

Option parsing happens too late to have any impact on the Rails
environment.  Rails accesses the environment name and memoizes it too
early in the boot process for a commandline option to have any impact on
the database connection, so we'll change this test to set the
environment from an environment variable (and ensure it still works when
running tests with `ruby`)

kaspth reviewed (commented):
❤️
